astacus/coordinator/plugins/clickhouse/steps.py

```diff
@@ -711,6 +747,60 @@ async def check_function_count(client: ClickHouseClient) -> bool:
         )
 
 
+@dataclasses.dataclass
+class RestoreKeeperMapTableDataStep(Step[None]):
+    zookeeper_client: ZooKeeperClient
+    keeper_map_path_prefix: str | None
+    clients: Sequence[ClickHouseClient]
+    sync_keeper_map_data_timeout: float
+
+    async def run_step(self, cluster: Cluster, context: StepsContext) -> None:
+        if self.keeper_map_path_prefix is None:
+            return
+
+        clickhouse_manifest = context.get_result(ClickHouseManifestStep)
+
+        if not clickhouse_manifest.keeper_map_tables:
+            return
+
+        data_path = None
+        row = None
+        async with self.zookeeper_client.connect() as connection:
+            transaction = connection.transaction()
+            for table in clickhouse_manifest.keeper_map_tables:
+                if not table.data:
+                    continue
+
+                data_path = os.path.join(self.keeper_map_path_prefix, table.name, "data")
+                if not await connection.exists(data_path):
+                    raise StepFailedError(f"KeeperMap table data path {table.name}/data doesn't exist in ZooKeeper")
+                for row in table.data:
+                    row_path = os.path.join(data_path, row.key)
+                    transaction.create(row_path, row.value)
+            await transaction.commit()
+
+        if data_path is not None and row is not None:
+            # Wait for final restored row to be visible on all nodes.
+            # Note that each ClickHouse instance maintains a single zookeeper
+            # session so reading from the system table will see the same data as
+            # reading from the KeeperMap table.
+            async def check_row_existence(client: ClickHouseClient) -> bool:
+                escaped_data_path = escape_sql_string(data_path.encode())
+                escaped_name = escape_sql_string(row.key.encode())
+                query = f"SELECT count(*) FROM system.zookeeper WHERE path = {escaped_data_path} AND name = {escaped_name}"
```

Comment:
My (limited) understanding is that ZK is _not_ read after write.  Writes go through the leader but not every node will receive the write immediately.  This ensures that all the ZK sessions that clickhouse is using have this node.

True this might be less efficient than expected, I will check the implementation of `system.zookeeper`.  It should be hopefully just a single `exists` query.